internal/jobservice/repository/sql_job_service.go

```diff
@@ -0,0 +1,266 @@
+	if rowsAffected == 0 {
+		return s.deleteJobsFromInMemoryMap(queue, jobSet)
+	}
+	// If we were able to delete from DB than we assume all is good
+	return rowsAffected, errDelete
+}
+
+func (s *SQLJobService) deleteJobsFromInMemoryMap(queue string, jobSet string) (int64, error) {
+	var rowsAffected int64 = 0
+	s.jobStatus.jobLock.Lock()
+	defer s.jobStatus.jobLock.Unlock()
+	for _, val := range s.jobStatus.jobMap {
+		if val.queue == queue && val.jobSetId == jobSet {
+			rowsAffected++
+			delete(s.jobStatus.jobMap, val.jobId)
+		}
+	}
+	return rowsAffected, nil
+}
+
+// Checks JobSet table to make determine if we should unsubscribe from JobSet
+// configTimeWithoutUpdates is a configurable value that is read from the config
+// We allow unsubscribing if the jobset hasn't been updated in configTime
+func (s *SQLJobService) CheckToUnSubscribe(queue string, jobSet string, configTimeWithoutUpdates int64) bool {
+	if !s.IsJobSetSubscribed(queue, jobSet) {
+		return false
+	}
+	currentTime := time.Now().Unix()
+	s.jobStatus.subscribeLock.RLock()
+	defer s.jobStatus.subscribeLock.RUnlock()
+	for _, val := range s.jobStatus.subscribeMap {
+		if val.queue == queue && val.jobSet == jobSet {
+			if (currentTime - val.lastRequestTimeStamp) > configTimeWithoutUpdates {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// Update JobSet Map with time that a Job in that JobSet was requested
+func (s *SQLJobService) UpdateJobSetTime(queue string, jobSet string) error {
+	s.jobStatus.subscribeLock.Lock()
+	defer s.jobStatus.subscribeLock.Unlock()
+	primaryKey := queue + jobSet
+	_, ok := s.jobStatus.subscribeMap[primaryKey]
+	if ok {
+		s.jobStatus.subscribeMap[primaryKey] = NewSubscribeTable(queue, jobSet)
+		return nil
+	} else {
+		return fmt.Errorf("queue %s jobSet %s is already unsubscribed", queue, jobSet)
+	}
+}
+
+// Delete Jobs in the map.
+func (s *SQLJobService) DeleteJobsInJobSet(queue string, jobSet string) (int64, error) {
+	result, err := s.db.Exec("DELETE FROM jobservice WHERE Queue=? AND JobSetId=?", queue, jobSet)
+	if err != nil {
+		return 0, err
+	}
+	return result.RowsAffected()
+}
+
+// Get a list of SubscribedJobSets (Queue JobSet)
+func (s *SQLJobService) GetSubscribedJobSets() []SubscribedTuple {
+	var returnJobSets []SubscribedTuple
+	s.jobStatus.subscribeLock.RLock()
+	defer s.jobStatus.subscribeLock.RUnlock()
+	for _, value := range s.jobStatus.subscribeMap {
+		tuple := &SubscribedTuple{Queue: value.queue, JobSet: value.jobSet}
+		returnJobSets = append(returnJobSets, *tuple)
+	}
+	return returnJobSets
+}
+
+// Save our in memory map to Database and delete from in memory map.
+func (s *SQLJobService) PersistDataToDatabase() error {
+	log.Info("Saving Data to Database")
+	s.jobStatus.jobLock.Lock()
+	defer s.jobStatus.jobLock.Unlock()
+	stmt, err := s.db.Prepare("INSERT INTO jobservice VALUES (?, ?, ?, ?, ?, ?)")
+	if err != nil {
+		panic(err)
+	}
+	for key, value := range s.jobStatus.jobMap {
```

Comment:
I think this will be quite slow because you'll have to persist every job to the database.  Also will this work if the job is already in the DB? In most DBs this would have to be an upsert rather than an insert but I don't know about SQLLite

Reply:
I'll add a unit test for that!